### PR TITLE
Add method for decoding sort keys, and use this in min/max for arbitrary types

### PR DESCRIPTION
**Pull request by Mytherin**
https://github.com/duckdb/duckdb/pull/10321 added the `create_sort_key` method, which allows us to take any input type, and create a binary-comparable `BLOB` according to a set of ordering constraints (e.g. `ASCENDING/DESCENDING/NULLS FIRST/NULLS LAST`). This PR adds the inverse - `DecodeSortKey` - albeit not as a scalar function but only for internal usage. This function allows us to take any binary blob created by the `create_sort_key` method, and reconstruct the original values that were ingested.

This method has multiple uses - but for this PR the primary use case is for this combination of functions to work as a fallback mechanism for types we do not specialize on. As DuckDB supports arbitrarily composible types (in the form of lists, structs, maps, that can all be nested) - we cannot specialize on every type - yet there are functions that need to work on every type (e.g. `MIN/MAX`). By using the sort keys, we only need to implement these methods for the `BLOB` type - after which we can support all types through e.g. `decode_sort_key(FUNCTION(create_sort_key(...)))`. This allows us to more easily add support for all types in a reasonably efficient manner (although not as efficient as specializing directly of course).

In this PR we switch the fallback implementation of `MIN/MAX` to use the new sort keys. The old fallback implementation constructed a `Vector` in the aggregate state that would hold only a single element, and used vector operations (e.g. `VectorOperations::Copy`) to assign elements to the aggregate. The sorting key implementation is significantly more performant - as the vector methods are not designed to operate on individual elements. In addition, this fallback method is also far more memory efficient, particularly when running the min/max operations on many groups, as vectors are designed to hold large arrays. 

Below are some benchmarks running min/max over structs/lists. We can see the large increase in performance, especially as the number of groups increases. The old implementation would also be killed by the OOM killer when running with many groups - which the new implementation does not suffer from.

###### Ungrouped
```sql
CREATE TABLE structs AS SELECT {'i': i} s FROM range(100000000) t(i);
SELECT MIN(s), MAX(s) FROM structs;

CREATE TABLE lists AS SELECT [i, i + 1, i + 2] l FROM range(100000000) t(i);
SELECT MIN(l), MAX(l) FROM lists;
```
|  Type   | v1.0.0 |  New  |
|---------|--------|-------|
| Structs | 1.50s  | 0.44s |
| Lists   | 1.64s  | 1.48s |

###### 10K Groups

```sql
CREATE TABLE structs AS SELECT i%10000 AS grp, {'i': i} s FROM range(100000000) t(i);
SELECT grp, MIN(s), MAX(s) FROM structs GROUP BY grp;

CREATE TABLE lists AS SELECT i%10000 AS grp, [i, i + 1, i + 2] l FROM range(100000000) t(i);
SELECT grp, MIN(l), MAX(l) FROM lists GROUP BY grp;
```

|  Type   | v1.0.0 |  New  |
|---------|--------|-------|
| Structs | 4.4s   | 0.44s |
| Lists   | 9.4s   | 2.1s  |

###### 10M Groups

```sql
CREATE TABLE structs AS SELECT i%10000000 AS grp, {'i': i} s FROM range(100000000) t(i);
SELECT grp, MIN(s), MAX(s) FROM structs GROUP BY grp;

CREATE TABLE lists AS SELECT i%10000000 AS grp, [i, i + 1, i + 2] l FROM range(100000000) t(i);
SELECT grp, MIN(l), MAX(l) FROM lists GROUP BY grp;
```

|  Type   | v1.0.0 | New  |
|---------|--------|------|
| Structs | KILLED | 2.0s |
| Lists   | KILLED | 64s  |

CC @lnkuiper @hawkfish

### CODE REVIEW

**Review by hawkfish — COMMENTED** [review]
Looks good, but had to ask about TIME_TZ.

File: src/core_functions/scalar/blob/create_sort_key.cpp [hunk 1]
```diff
@@ -642,40 +691,259 @@ static void FinalizeSortData(Vector &result, idx_t size) {
...
+		TemplatedDecodeSortKey<SortKeyConstantOperator<int16_t>>(decode_data, vector_data, result, result_idx);
+		break;
+	case PhysicalType::UINT32:
+		TemplatedDecodeSortKey<SortKeyConstantOperator<uint32_t>>(decode_data, vector_data, result, result_idx);
+		break;
+	case PhysicalType::INT32:
+		TemplatedDecodeSortKey<SortKeyConstantOperator<int32_t>>(decode_data, vector_data, result, result_idx);
+		break;
+	case PhysicalType::UINT64:
+		TemplatedDecodeSortKey<SortKeyConstantOperator<uint64_t>>(decode_data, vector_data, result, result_idx);
```
[the remark this answers is not in the document]
Comment (Mytherin) [reply src/core_functions/scalar/blob/create_sort_key.cpp:899]: That should be taken care of by the pluggable collations at a higher level - see https://github.com/duckdb/duckdb/pull/12492

[the remark this answers is not in the document]
Comment (Mytherin) [reply src/core_functions/scalar/blob/create_sort_key.cpp:899]: I suppose we can special case it and throw an exception if it ever gets thrown into `create_sort_key`

File: src/core_functions/scalar/blob/create_sort_key.cpp [hunk 2]
```diff
@@ -642,40 +691,259 @@ static void FinalizeSortData(Vector &result, idx_t size) {
...
 		ConstructSortKey(*sort_key_data[c], info);
 	}
-	FinalizeSortData(result, args.size());
+	FinalizeSortData(result, row_count);
+}
+
+void CreateSortKeyHelpers::CreateSortKey(Vector &input, idx_t input_count, OrderModifiers order_modifier,
+                                         Vector &result) {
+	// prepare the sort key data
+	vector<OrderModifiers> modifiers {order_modifier};
```
[the remark this answers is not in the document]
Comment (Mytherin) [reply src/core_functions/scalar/blob/create_sort_key.cpp:721]: The original create sort key was designed to take an arbitrary number of vectors - this is just to massage it into the right format for the rest of the code

File: src/core_functions/scalar/blob/create_sort_key.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mytherin) [reply src/core_functions/scalar/blob/create_sort_key.cpp:899]: But then that conflicts with some other use cases I have in mind for sort keys - which is just serializing/deserializing arbitrary types as blobs (with comparability not being of any concern) - so I will just leave this as-is

**Pull request merged**